### PR TITLE
ARROW-7610: [Java] Finish support for 64 bit int allocations

### PR DESCRIPTION
**Pull request by liyafan82**
Modified the `NettyAllocationManager` to allow for 64 bit allocations. If the requested size is > `Int.MAX_VALUE`, memory will be allocated using `PlatformDependent.allocateMemory(long)` otherwise the existing allocation method remains unchanged.

Unit tests were added that uses a custom AllocationManager with a smaller cutoff value to be able to test the new functionality within our current CI memory limits. Additional tests were added that allocate large buffers, but are required to run manually.

### CODE REVIEW

**Comment by liyafan82** [comment]
Some limitations of this PR:
1. We do not support large buffers in varchar and var binary vectors, as the offset width is fixed at 4. (I have noticed that ARROW-6111 has been opened to track this, and I would like to help if necessary).
2. We do not support E2E sending/receiving large buffers through IPC. This is because the current Flight implementation is relying on data structurs which are based on 32-bit integers (e.g. NettyArrowBuf#length and ByteArrayOutputStream#count are both based on int32). Fixing this may require non-trivial effort, and we may do it in another issue (if this is needed).

**Comment by emkornfield** [comment]
@liyafan82 I removed myself as assignee for ARROW-6111 feel free to pick it up if you have time.

File: java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java]: would it be simpler to make TYPE_WIDTH long for here an other places?  Are there downsides to that?

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java]:
```suggestion
 * the data buffer.
```

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java]: @jacques-n thoughts on making the default Unsafe to avoid any dependencies on Netty?

**Comment by emkornfield** [comment]
@liyafan82 can you elaborate on how flight impacts the normal IPC format? (perhaps a follow-up JIRA explaining it?)

**Comment by liyafan82** [comment]
> @liyafan82 I removed myself as assignee for ARROW-6111 feel free to pick it up if you have time.

Sure. I will try to proivide an implementation.

File: java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java]: Making it a long is a reasonable choice. 
I am not sure if is an overkill as TYPE_WIDTH is often a small number, and often used in scenarios where a small number is expected.

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (liyafan82) [reply java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java]: Nice catch. Thank you.

**Comment by liyafan82** [comment]
> @liyafan82 can you elaborate on how flight impacts the normal IPC format? (perhaps a follow-up JIRA explaining it?)

Sure. I will open a JIRA tracking it.

File: java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java
Comment (liyafan82) [reply java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java]: @emkornfield After more thoughts, I kind of prefer leaving the offset width int32. This makes sure we promote the related operations to int64 only when necessary. 

That being said, I will follow your advice and make the offset width int 64, if you have a reason, or feel strong about it.

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/UInt2Vector.java]: sounds ok to me.

**Comment by emkornfield** [comment]
> > @liyafan82 can you elaborate on how flight impacts the normal IPC format? (perhaps a follow-up JIRA explaining it?)

> 
> Sure. I will open a JIRA tracking it.

Did you get a chance, I'd like to understand what is left?

**Comment by liyafan82** [comment]
> > > @liyafan82 can you elaborate on how flight impacts the normal IPC format? (perhaps a follow-up JIRA explaining it?)
> 
> > Sure. I will open a JIRA tracking it.
> 
> Did you get a chance, I'd like to understand what is left?

Sure. I opened https://issues.apache.org/jira/browse/ARROW-7746 to track it. 
If the description is not clear enough, please let me know, and I will revise it.
The general idea is that some internal data structures are based on int32, so if the buffer size exceeds 2GB, an integer overflow will happen.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java]: @emkornfield I have changed the code to make the default unsafe, so that we support large buffers by default.

**Comment by liyafan82** [comment]
@emkornfield Could you please triage this PR to a committer, as it is blocking another issue (https://github.com/apache/arrow/pull/6782#issuecomment-606986823)?

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java:92]: Warn might be too hi, I think info is probably more appropriate.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java:67]: it seems like these should match up with the enums exactly? or probably do an non case sensitive comparison.

**Comment by emkornfield** [comment]
A couple of small comments. Sorry to flip-flop but lets keep the default allocator as Netty and then change it in the next release, when some of the other Memory PRs have been committed?  Can you also add details on how to configure the new allocator to the PR description so it gets put into release notes (we should also make sure to make note of this).

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java]: Revised to info. Thank you.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java]: Sounds good. Revised the code to match up with the enum exactly. Thank you.

**Comment by liyafan82** [comment]
Thanks a lot for your effort.

> A couple of small comments. Sorry to flip-flop but lets keep the default allocator as Netty and then change it in the next release, when some of the other Memory PRs have been committed? 

Sounds good. The default allocator is reverted to Netty.

> Can you also add details on how to configure the new allocator to the PR description so it gets put into release notes (we should also make sure to make note of this).

Good suggestion. I have added details to the description. Thank you.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java:41]: what happens when logging during static initialization?

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (emkornfield) [inline java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java:35]: nit: LOGGER seems to be more common in the code base.

**Review by emkornfield — CHANGES_REQUESTED** [review]
Two comments around logging.  I would like to make sure we aren't going to cause problems for users of the library by ignoring log configuration since we are logging in a  static initializer.  The other is a small nit.  I think this can be merged after we resolve those.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java]: Sounds good. Revised. Thank you.

File: java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/DefaultAllocationManagerOption.java:41]: If I understand your question correctly. You are concerned that when calling getDefaultAllocationManagerFactory, the logger may not have been initialized. If so, a NullPoinerException will be thrown. 

There is no need to worry about that. According to the Java specification (https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.4.2), static members in a class are initialized in textual order, as though they were a single block. 

In our code, initialization of the logger appears before the DEFAULT_ALLOCATION_MANAGER_FACTORY, so it always gets initialized first.  

In fact, this code path has been tested by all our test cases, as by default, no environmental variable or system property are specified when running the tests.

**Comment by liyafan82** [comment]
> Two comments around logging. I would like to make sure we aren't going to cause problems for users of the library by ignoring log configuration since we are logging in a static initializer. The other is a small nit. I think this can be merged after we resolve those.

Thanks a lot for your comments. Please see if I understand your question correctly.

**Comment by BryanCutler** [comment]
@liyafan82 I don't really see an answer to the question @jacques-n asked on the JIRA and mailing list:

>Per comments on the mailing list,
I think we should use the existing netty facades for this.
PlatformDependent.allocateMemory(long)
PlatformDependent.freeMemory(long)
These delegate to unsafe which also ensures that -XX:MaxDirectMemorySize is respected.
I don't think we need a new allocator. We can simply check the size and use this path if beyond an int in size within the existing allocator.

I agree with this that it is overkill to introduce different types of allocators if it can be handled with minimal changes to the current Netty based allocator. What are yours and @emkornfield thoughts on taking this approach?

**Comment by emkornfield** [comment]
I thought I replied on the mailing list to this but it would be nice to get rid of a hard dependency on netty for two reasons:
1.  Netty Version conflicts when consuming arrow 
2.  Setting the netty specific system parameter for later versions

There are other jiras which are also trying to trim the dependency on netty

**Comment by emkornfield** [comment]
Are there benefits to going through the netty facade/existing implementation?

**Comment by liyafan82** [comment]
> @liyafan82 I don't really see an answer to the question @jacques-n asked on the JIRA and mailing list:
> 
> > Per comments on the mailing list,
> > I think we should use the existing netty facades for this.
> > PlatformDependent.allocateMemory(long)
> > PlatformDependent.freeMemory(long)
> > These delegate to unsafe which also ensures that -XX:MaxDirectMemorySize is respected.
> > I don't think we need a new allocator. We can simply check the size and use this path if beyond an int in size within the existing allocator.
> 
> I agree with this that it is overkill to introduce different types of allocators if it can be handled with minimal changes to the current Netty based allocator. What are yours and @emkornfield thoughts on taking this approach?

@BryanCutler Thanks for your attention.
As @emkornfield suggested, we are trying to remove netty dependencies from the core of Arrow (Please see ARROW-4526 for details). A number of attempts are being made to separate netty and arrow implementations, and we want to finally move netty related stuffs to a separate module.

**Comment by liyafan82** [comment]
> Are there benefits to going through the netty facade/existing implementation?

One that comes to my mind is backward compatiblity?

**Comment by BryanCutler** [comment]
I'm all for removing dependencies on Netty, but that doesn't have to be done here. This could be done a lot easier by using the existing NettyAllocationManager instead of creating a new one, plus that also gives the benefit of using the default allocator that can also make large allocations. Adding an alternative allocation manager to Netty should probably be done as a separate issue with it's own discussion.

**Comment by emkornfield** [comment]
> I'm all for removing dependencies on Netty, but that doesn't have to be done here. This could be done a lot easier by using the existing NettyAllocationManager instead of creating a new one, plus that also gives the benefit of using the default allocator that can also make large allocations. Adding an alternative allocation manager to Netty should probably be done as a separate issue with it's own discussion.

I'm OK with this.  @liyafan82 I think it is probably easy to move the new allocator and selection code to a different PR and make the changes directly to the netty allocator in this PR?  Does that sound reasonable to you?

**Comment by liyafan82** [comment]
> > I'm all for removing dependencies on Netty, but that doesn't have to be done here. This could be done a lot easier by using the existing NettyAllocationManager instead of creating a new one, plus that also gives the benefit of using the default allocator that can also make large allocations. Adding an alternative allocation manager to Netty should probably be done as a separate issue with it's own discussion.
> 
> I'm OK with this. @liyafan82 I think it is probably easy to move the new allocator and selection code to a different PR and make the changes directly to the netty allocator in this PR? Does that sound reasonable to you?

@emkornfield Sounds reasonable to me. Thanks for the good suggestion @BryanCutler 
I will revise the code accordingly.

**Comment by liyafan82** [comment]
@BryanCutler @emkornfield 
I have revised the code for NettyAllocationManager to make it support allocating large buffers in this PR.
Another separate issue (ARROW-8481) is opened for the UnsafeAllocationManager. 

Please take a look. Thanks again for the good suggestion.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:42]: I don't think we should remove this which effectively replaces all allocations done in Arrow Java, which is a big change. `INNER_ALLOCATOR` also uses a pool which has some benefits. Instead, can you just change `requestedSize` to be a long, then check if it is over the max Int size and only then use `PlatformDependent.allocateMemory`?

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (BryanCutler) [inline java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:29]: please update

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (BryanCutler) [inline java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:65]: Can this be converted to a standard unit test now?

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java:34]: update

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java:181]: same here about making these standard unit tests

**Review by BryanCutler — CHANGES_REQUESTED** [review]

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:42]: Revised. Thank you for the good suggestion.

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java]: Nice catch. Thank you.

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:58]: Sounds good to me. 
The problem is that we set arrow.vector.max_allocation_bytes to 1048576 for every test case (to avoid OOM).  Please see the pom.xml file. 

So if we convert it to a test case, we cannot allocate too much memory.

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (liyafan82) [reply java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java]: Revised. Thank you.

File: java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java
Comment (liyafan82) [reply java/vector/src/test/java/org/apache/arrow/vector/TestLargeVector.java:175]: As discussed above, if we make it a unit test, we cannot allocate too much memory.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:47]: could you keep `allocatedSize = memoryChunk.capacity()` for this case?

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:43]: use `this.memoryChunk` to be consistent

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (BryanCutler) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:58]: What do you think about adding an additional constructor for `NettyAllocationManager` that also takes a cut-off value that will determine if `PlatformDependent.allocateMemory(requestedSize)` is used with a default of `Int.MAX_VALUE`? Then you could create a new `NettyAllocationManager` for testing with a small cut-off value that can be used to run normal tests.

**Review by BryanCutler — CHANGES_REQUESTED** [review]

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:50]: Sounds good. Reverted this accordingly.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Sure. Revised. Please check.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Revised accordingly. Thank you.

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:58]: Sounds good. 
I have revised the code to make the cut-off value configurable, and added cases to test the scenarios when the request size is below/above the cut-off value. Please see if it looks good to you. Thanks.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:63]: This should be an int because it can't go above `Int.MAX_VALUE`

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:106]: just check if `memoryChunk` is null here

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (BryanCutler) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:58]: So this is not run as part of the normal test suite correct? Does it need to run manually? if so could you put that in the javadoc

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:51]: none of this is necessary, please remove. you can create a custom `NettyAllocationManager` for tests and specify a different value there. See https://github.com/apache/arrow/blob/master/java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java#L393

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:96]: why add this? just return `memoryChunk`, it with either be null or a value

File: java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java:81]: misspelled `wil`

File: java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java
Comment (BryanCutler) [inline java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java:78]: how about set this to `NettyAllocationManager.DEFAULT_ALLOCATION_CUTOFF_VALUE` + someval

**Review by BryanCutler — CHANGES_REQUESTED** [review]

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Changed to int. Thanks for the good suggestion.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Revised accordingly. Thank you.

File: java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestLargeArrowBuf.java:62]: Yes, it needs to run manually. 
I have updated the javadoc accordingly.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java:39]: Removed. Thank you for the good suggestion.

File: java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/main/java/org/apache/arrow/memory/NettyAllocationManager.java]: Sorry. This is not needed. Reverted.

File: java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java]: Sorry about the typo. This comment is removed.

File: java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java
Comment (liyafan82) [reply java/memory/src/test/java/org/apache/arrow/memory/TestNettyAllocationManager.java]: Sounds good. Revised accordingly.

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
test failure is unrelated, merged to master. Thanks @liyafan82 !

**Comment by liyafan82** [comment]
@BryanCutler @emkornfield Thanks a lot for your effort.

**Pull request closed without merging**
